typed_python/Instance.hpp

```diff
@@ -0,0 +1,166 @@
+#pragma once
+
+#include <atomic>
+#include <string>
+// TODO #include "Type.hpp"
+
+class DeserializationBuffer;
+
+typedef uint8_t* instance_ptr;
+
+
+class Instance {
+private:
+    class layout {
+    public:
+        std::atomic<int64_t> refcount;
+        Type* type;
+        uint8_t data[];
+    };
+
+    Instance(layout* l) :
+        mLayout(l)
+    {
+    }
+
+    static layout* allocateNoneLayout() {
+        layout* result = (layout*)malloc(sizeof(layout));
+        result->refcount = 0;
+        result->type = None::Make();
+
+        return result;
+    }
+
+    static layout* noneLayout() {
+        static layout* noneLayout = allocateNoneLayout();
+        noneLayout->refcount++;
+
+        return noneLayout;
+    }
+
+public:
+    static Instance deserialized(Type* t, DeserializationBuffer& buf) {
+        t->assertForwardsResolved();
+
+        return createAndInitialize(t, [&](instance_ptr tgt) {
+            t->deserialize(tgt, buf);
+        });
+    }
+
+    static Instance create(Type*t, instance_ptr data) {
+        t->assertForwardsResolved();
+
+        return createAndInitialize(t, [&](instance_ptr tgt) {
+            t->copy_constructor(tgt, data);
+        });
+    }
+
+    template<class initializer_type>
+    static Instance createAndInitialize(Type* t, const initializer_type& initFun) {
+        return Instance(t, initFun);
+    }
+
+    Instance() {
+        //by default, None
+        mLayout = noneLayout();
+        mLayout->refcount++;
+    }
+
+    Instance(const Instance& other) : mLayout(other.mLayout) {
+        mLayout->refcount++;
+    }
+
+    Instance(instance_ptr p, Type* t) : mLayout(nullptr) {
+        t->assertForwardsResolved();
+
+        layout* l = (layout*)malloc(sizeof(layout) + t->bytecount());
+
+        try {
+            t->copy_constructor(l->data, p);
+        } catch(...) {
+            free(l);
+            throw;
+        }
+
+        l->refcount = 1;
+        l->type = t;
+
+        mLayout = l;
+    }
+
+    template<class initializer_type>
+    Instance(Type* t, const initializer_type& initFun) : mLayout(nullptr) {
+        t->assertForwardsResolved();
```

Comment:
not sure what to do with this constructor. If I put it in the `cpp` file, `_types.cc` which calls it will not find the correctly instantiated template. If I keep it in the `hpp` file, we cannot just have the forward declaration for `class Type`. We can look at this when you're back.